### ISSUE TITLE
New rule: no-implicit-module

### ISSUE THREAD

**Issue by thuba** [eslint/eslint#9429]
**Please describe what the rule should do:**

Disallow module scope variable declaration (variable that are not declared in function or Object).

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[X] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

_This should warn:_
> abc.js
```js
var a = 0;

function test () {
  return a + 1;
}
```

_While this should be OK:_
> efg.js
```js
function test(a) {
  return a + 1;
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**

Like `no-implicit-globals` disallow global variable and function declaration, it could be a good practice to disallow global variable inside a module scope (global variable are often a bad practice, not only in JS)

**Comment by not-an-aardvark** [comment]
Thanks for the proposal. I'm a bit confused about the purpose of the proposed rule, because the semantics of variable declarations are different in modules.

The `no-implicit-globals` rule exists because `var a = 0` in a script implicitly creates a property on the global object, which might not be intended. On the other hand, `var a = 0` does not create a property on the global object when used in a module, so there isn't any "implicit" behavior going on here.

Is the goal of the rule to simply disallow top-level variable declarations in modules?

**Comment by thuba** [comment]
Thanks for your response.

> Is the goal of the rule to simply disallow top-level variable declarations in modules?

Yes, that's exactly what I'm looking for, no top-level variable declaration in module, constant is OK (allowed/disallowed constant via a boolean option for example). The goal is to avoid multiple reading/writing to top level variable, which can be a source of error.

**Comment by not-an-aardvark** [comment]
I see -- so if I'm understanding correctly, the goal would be to avoid global mutable state in modules.

I have a few thoughts on the proposal:

* The rule would report code like this, where mutable state is used as an implementation detail (e.g. for caching) without being exposed as part of the module's API.

    ```js
    let foo;
    
    export function getFoo() {
        if (!foo) {
            // lazy load
            foo = expensiveComputation();
        }
        return foo;
    }
    ```

    I agree that it's almost always a bad idea to *expose* global mutable state. I don't think it's necessarily a problem to *have* global mutable state, provided that the state is an implementation detail and the module's API acts as if it were immutable. The proposed rule would report the latter, even if the goal is to prevent the former.

* It's worth noting that module-scoped variables are not the only way to accidentally expose mutable state. For example, a user could modify a property of a module-scoped variable, and it wouldn't be caught by this rule.
* Bikeshed: I'm not a fan of the name `no-implicit-module`. I assume `no-implicit-module` was a modification of the existing `no-implicit-global` rule, but I don't think that wording makes sense (there isn't anything implicit going on, and the module itself isn't the problem.) Maybe something like `no-global-module-bindings` or something would be better.

**Comment by thuba** [comment]
- No problem to change the name, I found it in five minutes and know that is not the better choice. So, `no-mutable-module-bindings` would be better.

- Based on your example, to avoid potential error:

```js
export default class Bar {

  constructor() {
    this.foo;
  }

  get foo() {
    if (!this.foo) {
      this.foo = expensiveComputation();
    }
    return this.foo;
  }

}
```
`singleton` must be in class' scope, not directly in module scope. Furthermore, I's not always the [best design pattern choice](https://www.michaelsafyan.com/tech/design/patterns/singleton)

**Comment by not-an-aardvark** [comment]
There are some cases where storing the state on a class like you've described doesn't make sense. Here's a more concrete example:

```js
let dependency;

export async function rarelyCalledFunction() {
  if (!dependency) {
    dependency = await import('rarely-needed-module');
  }

  dependency.doSomething();
}
```

Module objects are always singletons, so this type of code would make sense to lazy-load a module which is rarely used.

**Comment by thuba** [comment]
Ok, I understand, thank you for explanation!
